### PR TITLE
JBPM-9924- Define a default value of the process variable at design time

### PR DESCRIPTION
**Pull request by VaniHaripriya**
JBPM-9924- https://issues.redhat.com/browse/JBPM-9924

### CODE REVIEW

File: jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java [hunk 1]
```diff
@@ -173,6 +173,7 @@ public void enforceRequiredVariables() {
         VariableScope variableScope = getVariableScope();
         for (Variable variable : variableScope.getVariables()) {
             String name = variable.getName();
+			Object defaultValue = variable.getMetaData("defaultValue");
```
Comment (elguardian) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java:176]: this will cause the check to be triggered before tha default value is set, therefore causing an error when default value is provided (which is not coherent)

**Comment by elguardian** [comment]
please keep in mind that the proper place to do this is not here but the process instance factory where parameters are set and not here.

**Review by elguardian — CHANGES_REQUESTED** [review]
this requires some rework

File: jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java (hunk 1 above)
Comment (elguardian) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java:176]: removed

File: jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java [hunk 2]
```diff
@@ -186,6 +187,9 @@ public void enforceRequiredVariables() {
                 }
                 
             }
+			if (defaultValue != null) {
```
Comment (elguardian) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java:190]: this needs to be removed.

File: jbpm-flow/src/main/java/org/jbpm/process/instance/AbstractProcessInstanceFactory.java [hunk 3]
```diff
@@ -52,6 +54,16 @@ public ProcessInstance createProcessInstance(Process process, CorrelationKey cor
         }
         // TODO: should be part of processInstanceImpl?
         VariableScope variableScope = (VariableScope) ((ContextContainer) process).getDefaultContext( VariableScope.VARIABLE_SCOPE );
+		if (variableScope != null) {
+			for (Variable variable : variableScope.getVariables()) {
+				DataType dataType = variable.getType();
+				Object defaultValue = variable.getMetaData("defaultValue");
+				if (defaultValue != null && defaultValue instanceof String && variable.getValue() == null) {
+					defaultValue = dataType.readValue((String) defaultValue);
+					variable.setValue(defaultValue);
```
Comment (elguardian) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/AbstractProcessInstanceFactory.java:63]: the variable scope is the "model" ... yo need to set the "instance"

if(varaibleScopeInstance.getVariable() == null) {
   varaibleScopeInstance.setVariable(name, defaultValue)
}

**Review by elguardian — CHANGES_REQUESTED** [review]

File: jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java [hunk 4]
```diff
@@ -171,22 +171,25 @@ public void setContextInstanceContainer(ContextInstanceContainer contextInstance
...
-                        
-                    }
-                    // otherwise check variables                    
-                } else if (!hasData(variables.get(name))) {
-                    throw new VariableViolationException(getProcessInstance().getId(), name, "Variable '" + name + "' is required but not set");
-                }
-                
-            }
-        }
+		for (Variable variable : variableScope.getVariables()) {
```
Comment (elguardian) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java:174]: try to avoid unnecessary changes like this one. otherwise PR gets difficult to track.

**Review by elguardian — APPROVED** [review]
LGTM waiting QE to check on this.

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 5]
```diff
@@ -162,4 +166,37 @@ public void testMultiValuedKeyUniqueButInclusive() {
         ProcessInstance processInstance2 = ksession.startProcess(PROCESS, key2, null);
         assertProcessInstanceActive(processInstance2.getId());
     }
+    
+    @Test
+    public void testCreateProcessInstanceDefaultValue() {
+        CorrelationKey key = keyFactory.newCorrelationKey(SIMPLE_KEY);
+
+        Map<String, Object> parameters = new HashMap<String, Object>();
```
Comment (fjtirado) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:174]: Collections.emptyMap()

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 6]
```diff
@@ -162,4 +166,37 @@ public void testMultiValuedKeyUniqueButInclusive() {
...
+    @Test
+    public void testCreateProcessInstanceComplexTypeDefaultValue() {
+       	String address = "<org.jbpm.test.domain.Address>\n"
+    			+ "          	   <street>abc</street>\n"
+    			+ "						   <number>29</number>\n"
+    			+ "						   <city>def</city>\n"
+    			+ "						</org.jbpm.test.domain.Address>";
+        CorrelationKey key = keyFactory.newCorrelationKey(SIMPLE_KEY);
+
+        Map<String, Object> parameters = new HashMap<String, Object>();
```
Comment (fjtirado) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:192]: Collections.emptyMap

**Review by fjtirado — COMMENTED** [review]
How about a test checking that if variable value is provided as parameter and default value is provided in bpmn file, value provided as parameter is taken. 
Also, this is for @elguardian, should default vale trigger a variable value changed event? My guess is yes, but Im unsure

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 7]
```diff
@@ -162,4 +166,37 @@ public void testMultiValuedKeyUniqueButInclusive() {
...
+        assertEquals(Integer.valueOf(1), processInstance.getVariable("secondVar"));
+        
+    }    
+    
+    @Test
+    public void testCreateProcessInstanceComplexTypeDefaultValue() {
+       	String address = "<org.jbpm.test.domain.Address>\n"
+    			+ "          	   <street>abc</street>\n"
+    			+ "						   <number>29</number>\n"
+    			+ "						   <city>def</city>\n"
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:188]: Test failing because blanks shouldn't be present after tabs.
Anyway, I think this will also fail in windows due to a different separator. Better to use "xmlunit" library for this kind of comparison.

**Review by gmunozfe — CHANGES_REQUESTED** [review]

File: jbpm-flow/src/main/java/org/jbpm/process/instance/AbstractProcessInstanceFactory.java [hunk 8]
```diff
@@ -20,7 +20,9 @@
 
 import org.drools.core.common.InternalKnowledgeRuntime;
 import org.jbpm.process.core.ContextContainer;
+import org.jbpm.process.core.context.variable.Variable;
 import org.jbpm.process.core.context.variable.VariableScope;
+import org.jbpm.process.core.datatype.DataType;
```
Comment (gmunozfe) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/AbstractProcessInstanceFactory.java:25]: This is not used, please remove

File: jbpm-flow/src/main/java/org/jbpm/process/instance/AbstractProcessInstanceFactory.java [hunk 9]
```diff
@@ -64,6 +66,17 @@ public ProcessInstance createProcessInstance(Process process, CorrelationKey cor
...
+			for (Variable variable : variableScope.getVariables()) {
+				String name = variable.getName();
+				Object defaultValue = variable.getMetaData("defaultValue");
+				if (variableScopeInstance.getVariable(name) == null && defaultValue != null) {
+					variableScopeInstance.setVariable(name,
+							variableScope.validateVariable(process.getName(), name, defaultValue));
+
+				}
+			}
+		}
```
Comment (gmunozfe) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/AbstractProcessInstanceFactory.java:79]: Check indentation, avoid tabs in favour of spaces

**Comment by elguardian** [comment]
@fjtirado setting the variables at the beginning will set the same triggers as provided by the end user. so that should be ok.

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java (hunk 7 above)
Comment (VaniHaripriya) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java]: Sure, updated the test cases accordingly.

File: jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java (hunk 4 above)
Comment (gmunozfe) [reply jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java:174]: @VaniHaripriya could you undo these indentation changes as Enrique suggested?

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 10]
```diff
@@ -162,4 +168,55 @@ public void testMultiValuedKeyUniqueButInclusive() {
...
+        assertEquals(Integer.valueOf(1), processInstance.getVariable("secondVar"));
+
+    }
+
+    @Test
+    public void testCreateProcessInstanceComplexTypeDefaultValue() throws SAXException, IOException {
+
+        String address = "<org.jbpm.test.domain.Address>\n" + "          	 <street>abc</street>\n"
+                + "						   <number>29</number>\n" + "						   <city>def</city>\n"
+                + "						</org.jbpm.test.domain.Address>";
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:206]: I think you can write this without spaces and carriage returns (that's why xmlunit is useful)

**Review by gmunozfe — CHANGES_REQUESTED** [review]

**Comment by gmunozfe** [comment]
jenkins retest this

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 11]
```diff
@@ -162,4 +168,53 @@ public void testMultiValuedKeyUniqueButInclusive() {
...
+
+        CorrelationKey key = keyFactory.newCorrelationKey(SIMPLE_KEY);
+
+        Map<String, Object> parameters = new HashMap<String, Object>();
+
+        WorkflowProcessInstance processInstance = (WorkflowProcessInstance) ksession.startProcess(PROCESS, key,
+                parameters);
+
+        DetailedDiff diff = new DetailedDiff(
+                XMLUnit.compareXML(address, (String) processInstance.getVariable("complexVar")));
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:214]: It's not clear to me that this testing is working properly in the sense of returned variable is an instance of String. I was expecting to get a custom org.jbpm.test.domain.Address. For example, if the parameters map sets up a "complexVar" with an Address value, it's obtained again an object from this type, wdyt?

**Review by gmunozfe — CHANGES_REQUESTED** [review]
There are test failures due to readValue invocation for some types:
```
at org.jbpm.process.core.datatype.impl.type.IntegerDataType.readValue(IntegerDataType.java:55)
at org.jbpm.process.core.context.variable.VariableScope.validateVariable(VariableScope.java:111)
```

**Comment by gmunozfe** [comment]
Failures are unrelated, let's try again.
Jenkins retest this

File: jbpm-test-coverage/pom.xml [hunk 12]
```diff
@@ -125,6 +125,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>xmlunit</groupId>
+      <artifactId>xmlunit</artifactId>      
+      <scope>test</scope>
+    </dependency>
```
Comment (gmunozfe) [inline jbpm-test-coverage/pom.xml:132]: this is not longer used, you can remove it

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 13]
```diff
@@ -16,26 +16,35 @@
...
 package org.jbpm.test.functional;
 
+import static org.junit.Assert.assertEquals;
+import java.io.IOException;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
+import org.custommonkey.xmlunit.XMLUnit;
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:26]: Same

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 14]
```diff
@@ -16,26 +16,35 @@
...
 import org.assertj.core.api.Assertions;
+import org.custommonkey.xmlunit.DetailedDiff;
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:28]: Same

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 15]
```diff
@@ -16,26 +16,35 @@
...
+import org.jbpm.test.domain.Address;
 import org.junit.Test;
 import org.kie.api.runtime.manager.audit.VariableInstanceLog;
 import org.kie.api.runtime.process.ProcessInstance;
+import org.kie.api.runtime.process.WorkflowProcessInstance;
 import org.kie.internal.KieInternalServices;
 import org.kie.internal.process.CorrelationAwareProcessRuntime;
 import org.kie.internal.process.CorrelationKey;
 import org.kie.internal.process.CorrelationKeyFactory;
+import org.xml.sax.SAXException;
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:39]: Same

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java [hunk 16]
```diff
@@ -46,7 +55,7 @@
     @Override
     public void setUp() throws Exception {
         super.setUp();
-
+        XMLUnit.setIgnoreWhitespace(true);
```
Comment (gmunozfe) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/CorrelationKeyTest.java:58]: Same here

**Review by gmunozfe — APPROVED** [review]
Looks good to me, great work @VaniHaripriya !
Just a few comments regarding non-used dependencies, if you want to delete them.

File: jbpm-flow/src/main/java/org/jbpm/process/core/context/variable/VariableScope.java [hunk 17]
```diff
@@ -102,14 +102,17 @@ public Object validateVariable(String processName, String name, Object value) {
...
         } else if (value != null) {
             DataType type = var.getType();
+            ObjectDataType obj = new ObjectDataType();
             boolean isRightType = type == null || type.verifyDataType(value);
             if (!isRightType ) {
                 if (variableStrictEnabled) {
                     throw new IllegalArgumentException("Variable '" + name + "' has incorrect data type expected:" + var
                             .getType().getStringType() + " actual:" + value.getClass().getName());
-                } else if (value instanceof String) {
+                } else if (value instanceof String && (type.getClass() != obj.getClass())) {
```
Comment (gmunozfe) [inline jbpm-flow/src/main/java/org/jbpm/process/core/context/variable/VariableScope.java:111]: Probably you can write it more clear like this
```suggestion
                } else if (value instanceof String) {
                    if (type.getClass() == obj.getClass()) {
                        value = type.readValue(value.toString());
                    else {
                        value = type.valueOf(value.toString());
                    }
                }
```

**Review by gmunozfe — COMMENTED** [review]
Also a suggestion for better readibility

File: jbpm-test-coverage/pom.xml (hunk 12 above)
Comment (VaniHaripriya) [reply jbpm-test-coverage/pom.xml:132]: Sure Gonzalo, will make the changes accordingly.. Thank you!
